### ISSUE TITLE
Support cross-fading

### ISSUE THREAD

**Issue by Justkant** [DoubleSymmetry/react-native-track-player#670]
Hey, nice work on this library.

I saw this comment https://github.com/react-native-kit/react-native-track-player/issues/172#issuecomment-378645916 and was wondering, how hard would it be to implement the native transition ?

I don't think we would be able to support the transition while casting but it would be nice to have it while playing on the device.

If you think the implementation is possible, I can help by sending a PR. I would just need some tips as to where to start looking.
Thanks

**Comment by Guichaguri** [comment]
As the native libraries we use don't support crossfading yet, you'll have to keep two player instances running at the same time, updating their volume through a timer. You'll also have to implement logic to buffer the track before starting to crossfade, otherwise you'll get one track slowly fading out and nothing else playing.

That would require quite a bit of code to be refactored though.

**Comment by Justkant** [comment]
I already have this logic in JS with other players but could implement it natively I think.

You'll probably prefer that I fork the player as the code would change quite a bit.
I will benchmark other native player solutions for my use case and ping back here to tell you if I start working on this native cross-fading or not

**Comment by bepehr** [comment]
> As the native libraries we use don't support crossfading yet, you'll have to keep two player instances running at the same time, updating their volume through a timer. You'll also have to implement logic to buffer the track before starting to crossfade, otherwise you'll get one track slowly fading out and nothing else playing.
> 
> That would require quite a bit of code to be refactored though.

Can you provide an example ? I`m looking for fading song too

**Comment by chrismang** [comment]
The fade song feature is in progress or it doesn't being develop yet?

**Comment by Bickson** [comment]
@Justkant Did you solve your crossfade problem? I need to do the same, just looking for inspiration.
Thankful for all help I can get :)

**Comment by Bickson** [comment]
is it possible to initiate 2 instances of TrackPlayer ?

**Comment by Justkant** [comment]
> Did you solve your crossfade problem?

@Bickson No, I had no time to get back to our experiment with RN

**Comment by ODINA101** [comment]
any update?

**Comment by robmim** [comment]
A quick implementation of rampUp or fadeIn based on this code : https://github.com/miohtama/Krusovice/blob/master/src/tools/fade.js

`const targetVolume = 1;
  const rampTime = 6000;
  const tick = 50;

  const fadeIn = async () => {
    const volumeIncrease = targetVolume / (rampTime / tick);
    const volume = await TrackPlayer.getVolume();
    const vol = Math.min(targetVolume, volume + volumeIncrease);

    await TrackPlayer.setVolume(vol).then(() => {
      // Have we reached target volume level yet?
      if (volume < targetVolume) {
        // Keep up going until 11
        setTimeout(fadeIn, tick);
      }
    });
  };`
  
  And use it on onTrackItemPress function like this
  
  `const onTrackItemPress = async (track: TrackPlayer.Track) => {
    await TrackPlayer.stop();

    await TrackPlayer.setVolume(0).then(() => {
      fadeIn();
    });

    await TrackPlayer.reset();

    setSelectedTrack(track);
  };`

**Comment by bradleyflood** [comment]
Thanks all. I've added this as a Feature Request in the Project Backlog. We will review.

Cheers

**Comment by lovegaoshi** [comment]
im having mild success with crossfading.

the work is based on https://github.com/doublesymmetry/KotlinAudio/pull/92 which is exactly 2 rotating exoplayers. 

[with media3](https://github.com/lovegaoshi/react-native-track-player/tree/de6bb1ae52fe41144f258f64780093e56a5977f0)  its trickier, as notification and service management is (can/should be) handled by the lib itself. But the idea is the same. I opted to make the rotating exoplayers wrapped in the forwarding player, and expose the forwarding player to MediaSession.
